presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetPageSourceFactory.java

```diff
@@ -197,13 +201,29 @@ public static ParquetPageSource createParquetPageSource(
                     systemMemoryContext,
                     maxReadBlockSize);
 
-            return new ParquetPageSource(
-                    parquetReader,
-                    fileSchema,
-                    messageColumnIO,
-                    typeManager,
-                    columns,
-                    useParquetColumnNames);
+            ImmutableList.Builder<Type> prestoTypesBuilder = ImmutableList.builder();
+            ImmutableList.Builder<Optional<Field>> internalFieldsBuilder = ImmutableList.builder();
+            for (int columnIndex = 0; columnIndex < columns.size(); columnIndex++) {
+                HiveColumnHandle column = columns.get(columnIndex);
+                org.apache.parquet.schema.Type parquetField = parquetFields.get(columnIndex);
+
+                String name = column.getName();
+                Type prestoType = typeManager.getType(column.getTypeSignature());
+
+                prestoTypesBuilder.add(prestoType);
+
+                if (parquetField == null) {
+                    internalFieldsBuilder.add(Optional.empty());
+                }
+                else {
+                    String columnName = useParquetColumnNames ? name : fileSchema.getFields().get(column.getHiveColumnIndex()).getName();
+                    internalFieldsBuilder.add(constructField(prestoType, lookupColumnByName(messageColumnIO, columnName)));
+                }
+            }
+            List<Type> types = prestoTypesBuilder.build();
+            List<Optional<Field>> internalFields = internalFieldsBuilder.build();
```

Comment:
You can inline `types` and `internalFields`.
Then you can rename builders to `prestoTypes` and `internalFields` respectively.